include/drivers/pwm.h

```diff
@@ -89,6 +177,131 @@ static inline int z_impl_pwm_pin_set_cycles(const struct device *dev,
 	return api->pin_set(dev, pwm, period, pulse, flags);
 }
 
+#ifdef CONFIG_PWM_CAPTURE
```

Comment:
The `pwm_pin_configure_capture()` function _is_ included in the documentation since the documentation build sets `CONFIG_PWM_CAPTURE`.

Attempting to use PWM capture in an application without `CONFIG_PWM_CAPTURE` should result in a linker error (similar to what we discussed in an API meeting a while back regarding the ADC API after the introduction of b7ce84930794a41ebe2c735f1d8c54f94712c7ba).

Reply:
> The `pwm_pin_configure_capture()` function _is_ included in the documentation since the documentation build sets `CONFIG_PWM_CAPTURE`.

Right, I missed that.

> Attempting to use PWM capture in an application without `CONFIG_PWM_CAPTURE` should result in a linker error (similar to what we discussed in an API meeting a while back regarding the ADC API after the introduction of [b7ce849](https://github.com/zephyrproject-rtos/zephyr/commit/b7ce84930794a41ebe2c735f1d8c54f94712c7ba)).

I recall that discussion but I don't recall that there was a conclusion that such attempts should result in linker errors. This is not the case for `CONFIG_SPI_ASYNC` or `CONFIG_UART_ASYNC_API`, for example. And I'm actually not convinced that linker errors are the best way to signal such attempts. To ensure that the required option is enabled, one can use a corresponding dependency in Kconfig or a build assertion in the code that uses this API, or one can check the option with `IS_ENABLED()` before doing the actual call. And I can imagine a scenario, not very common but still, where such optional API could be called from some precompiled library capable of handling properly the `-ENOTSUP` error returned when the API is not available. The approach with linker errors enforces that the caller code must be compiled with the same setting of the option as the driver code, so one would need two versions of such precompiled library.
That's only my opinion, though. Others can see it differently, so I'm not going to try to block this PR for this reason.

Reply:
The consensus, as I recall, was that the `#ifdef` should be around the `z_impl_*` function, but that the syscall should not be `#ifdef`'ed out. This will produce a linker error if trying to use the syscall without the Kconfig enabled. Here, the function in question is not a syscall (only an implementation) thus the entire function should be `#ifdef`'ed.

I believe the scenario, you list, is highly theoretical and will run into other, similar problems in many other parts of the kernel code.

Reply:
@henrikbrixandersen You requested an opinion from Slack: Not familiar with the code but my general opinion is that it's best to avoid `#ifdef`s in the header files (basically the only prober usage of that in header files, IMO, is to optimize structs). 

If my application uses some function which isn't available due to a mismatch in the config, I'd rather get that message during compilation/linking, than at runtime, i.e. have a `#ifdef` around the implementation, rather than having the function return `-EOPNOTSUPP` (please don't use `-ENOTSUP` for this as that has a different meaning). There are some cases where the function may want to return `-EOPNOTSUPP` instead of being guarded by a `#ifdef` though. 

So my 2 cents:
1) Don't guard the function declaration
2) Guard the function implementation
2.a) If not possible/applicable to guard the function implementation, return `-EOPNOTSUPP`.

One compromise could be to mention the reliance of a specific KConfig in the function documentation, if you want.

Reply:
Regarding `-ENOTSUP` see #23727; it's wrong, but it's unlikely to change.  Also `EOPNOTSUP` appears to be Linux-specific, while `ENOTSUP` is POSIX; the two do have the same value, but I don't want to see documentation referencing a non-POSIX identifier in the `E*` namespace.

I too would rather get build/link errors to determine when something is missing.  But "consensus" in Zephyr depends on who was attending the telecon where it was discussed, and generally isn't recorded anywhere, so it's a weak concept.

Reply:
Looking back at the top comment per the slack request:  Where an API structure lacks the API pointer the wrapper should return `-ENOTSUP`, if other drivers with the same API may support the feature.  Existing practice of unconditionally invoking the pointer is perhaps unsafe.

But where an entire suite of functionality in all drivers using a particular subsystem driver API is disabled, I would rather discover use of the unsupported API as a link time error, as is currently done in this PR by eliding the function definition.  This seems more robust in the face of somebody changing the Kconfig to remove the support: I would want the application to fail if it required the support, and erroneous build asserts intended to check for this case presence may be incorrect or become outdated.

I do agree that the description of the conditionally-supported function should include a reference to the required Kconfig.  Based on existing use in `kernel.h` and [elsewhere](https://github.com/zephyrproject-rtos/zephyr/blob/6297c667c591055e6c2f88d8b9a2b01a05c9bb94/include/sys/cbprintf.h#L53) something like this seems reasonable:

```
 * @note @option{CONFIG_PWM_CAPTURE} must be selected for this function to be available.
```

The concern with linking to an external library that has different expectations is valid, but I don't think it's safe in Zephyr anyway. AFAIK we don't define a stable ABI.

Reply:
> Regarding `-ENOTSUP` see #23727; it's wrong, but it's unlikely to change. Also `EOPNOTSUP` appears to be Linux-specific, while `ENOTSUP` is POSIX; the two do have the same value, but I don't want to see documentation referencing a non-POSIX identifier in the `E*` namespace.
> 
> I too would rather get build/link errors to determine when something is missing. But "consensus" in Zephyr depends on who was attending the telecon where it was discussed, and generally isn't recorded anywhere, so it's a weak concept.

Thanks for that information! I was actually going by the definition of the error codes in `lib/libc/minimal/include/errno.h` in Zephyr:

```
#define ENOTSUP 35      /* Unsupported value */
#define EOPNOTSUPP 45      /* Operation not supported on socket */
```

I won't be commenting more about this in this threads (seems to be out of scope), but I can't find much information about whether or not `EOPNOTSUPP` is POSIX, except that most sources I find on Google seem to say that it is indeed POSIX and not Linux (see also https://pubs.opengroup.org/onlinepubs/9699919799/).

Reply:
> I do agree that the description of the conditionally-supported function should include a reference to the required Kconfig.

Thank you for the references. I have added a note to the conditional API functions about enabling the Kconfig for them to be available.

Reply:
As per todays API meeting, the current [API design guidelines for conditional APIs](https://docs.zephyrproject.org/latest/guides/design_guidelines.html#conditional-data-and-apis) states the following:

> Function declarations that are available only when the option is enabled should be provided unconditionally. Add a note in the description that the function is available only when the specified feature is enabled, referencing the required Kconfig symbol by name. (In cases where the function is used but not enabled a link-time error will be generated.)

The implementation in this PR is consistent with the above design guidelines.